x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetFlamegraphAction.java

```diff
@@ -0,0 +1,287 @@
+            builder.addSamplesExclusive(0, 0);
+
+            int frameCount = stackTrace.frameIds.size();
+            for (int i = 0; i < frameCount; i++) {
+                String frameId = stackTrace.frameIds.get(i);
+                String fileId = stackTrace.fileIds.get(i);
+                Integer frameType = stackTrace.typeIds.get(i);
+                Integer addressOrLine = stackTrace.addressOrLines.get(i);
+                StackFrame stackFrame = response.getStackFrames().getOrDefault(frameId, EMPTY_STACKFRAME);
+                String executable = response.getExecutables().getOrDefault(fileId, "");
+
+                for (Frame frame : stackFrame.frames()) {
+                    String frameGroupId = createFrameGroupId(fileId, addressOrLine, executable, frame.fileName(), frame.functionName());
+
+                    int nodeId;
+                    if (builder.isExists(frameGroupId)) {
+                        nodeId = builder.getNodeId(frameGroupId);
+                        builder.addSamplesInclusive(nodeId, samples);
+                    } else {
+                        nodeId = builder.addNode(
+                            fileId,
+                            frameType,
+                            frame.inline(),
+                            executable,
+                            addressOrLine,
+                            frame.functionName(),
+                            frame.functionOffset(),
+                            frame.fileName(),
+                            frame.lineNumber(),
+                            samples,
+                            frameGroupId
+                        );
+                    }
+                    if (i == frameCount - 1) {
+                        // Leaf frame: sum up counts for exclusive CPU.
+                        builder.addSamplesExclusive(nodeId, samples);
+                    }
+                    builder.setCurrentNode(nodeId);
+                }
+            }
+        }
+        return builder.build();
+    }
+
+    @SuppressForbidden(reason = "Using pathSeparator constant to extract the filename with low overhead")
+    private static String getFilename(String fullPath) {
+        if (fullPath == null || fullPath.isEmpty()) {
+            return fullPath;
+        }
+        int lastSeparatorIdx = fullPath.lastIndexOf(File.pathSeparator);
+        return lastSeparatorIdx == -1 ? fullPath : fullPath.substring(lastSeparatorIdx + 1);
+    }
+
+    private static String createFrameGroupId(
+        String fileId,
+        Integer addressOrLine,
+        String exeFilename,
+        String sourceFilename,
+        String functionName
+    ) {
+        StringBuilder sb = new StringBuilder();
+        if (functionName.isEmpty()) {
+            sb.append(fileId);
+            sb.append(addressOrLine);
+        } else {
+            sb.append(exeFilename);
+            sb.append(functionName);
+            sb.append(getFilename(sourceFilename));
+        }
```

Comment:
When I first compared the Kibana server flamegraph and this flamegraph, I noticed that the edges didn't match. I walked through the code and I suspect part of the reason is that the internal sorting used by `TreeMap` is different than the `localeCompare` used in Kibana.

However, as part of the comparison, I noticed that the serialization of the frame group ID was different than Kibana. Upon further inspection, I realize that the Kibana implementation has a bug that would require `createFrameGroupID` to be changed slightly. Instead of changing that here, I'll raise an issue to document it.